File: csharp/proverb/Proverb.cs
using System;

public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        if (subjects.Length == 0)
        {
            return Array.Empty<string>();
        }

        string[] result = new string[subjects.Length];

        for (int i = 0; i < subjects.Length; i++)
        {
            if (i == subjects.Length - 1)
            {
                result[i] = $"And all for the want of a {subjects[0]}.";
            }
            else
            {
                result[i] = $"For want of a {subjects[i]} the {subjects[i + 1]} was lost.";
            }
        }

        return result;
    }
}

File: java/proverb/src/main/java/Proverb.java
class Proverb {

    private final String[] words;

    Proverb(String[] words) {
        this.words = words;
    }

    String recite() {
        if (words.length == 0) {
            return "";
        }

        final var result = new StringBuilder();

        for (int i = 0; i < words.length; i++) {
            if (i == words.length - 1) {
                result.append("And all for the want of a %s.".formatted(words[0]));
            } else {
                result.append("For want of a %s the %s was lost.\n".formatted(words[i], words[i + 1]));
            }
        }

        return result.toString();
    }

}
